src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -945,9 +949,19 @@ private RegexNode ReduceAlternation()
 
                 default:
                     ReduceSingleLetterAndNestedAlternations();
-                    RegexNode node = ReplaceNodeIfUnnecessary(Nothing);
-                    node = ExtractCommonPrefixText(node);
-                    node = ExtractCommonPrefixOneNotoneSet(node);
+                    RegexNode node = ReplaceNodeIfUnnecessary();
+                    if (node.Type == Alternate)
+                    {
+                        node = ExtractCommonPrefixText(node);
+                        if (node.Type == Alternate)
+                        {
+                            node = ExtractCommonPrefixOneNotoneSet(node);
+                            if (node.Type == Alternate)
+                            {
+                                node = RemoveRedundantEmptiesAndNothings(node);
+                            }
+                        }
+                    }
```

Comment:
should Reduce() be called on the node at this point, or will it get subsequently?

Reply:
I don't think so?

We do potentially leave some optimizations on the table due to the last round of optimizations exposing additional opportunities and us not doing another pass.  You could imagine us looping over all of the optimizations until no more changes were applied.  But that would also add non-trivial expense.

I've noticed this in particular after we do our FinalOptimize pass, which walks down the right-hand side of the tree letting each node know it's at the end and thus able to be treated as atomic, and we don't then subsequently reapply other optimizations.  For example, imagine you had the expression `abc(?:d+?)`.  Because it's at the end of the expression, that `d+?` can be made atomic, and an atomic lazy can become a repeater of the loop's minimum bound, so `d+?` just becomes `d{1}` which is the same of course as `d`, which means the whole expression becomes `abcd`.  But at that point in our tree, we'll have a concatenation of a Multi ("abc") and a One ('d') rather than a single Multi ("abcd") because the optimization that combines those into a single node doesn't get run again.

Reply:
The jit has a mode where it will repeat its main optimization passes, but we don't use it for anything but stress testing.

As @stephentoub notes, iterating is likely costly and typically doesn't have great ROI. If you can anticipate what sorts of subsequent optimizations might happen, you can perhaps mitigate some of the cost by being more selective about when and where you try re-optimizing.

There rarely seems to be a well-formed notion of a universal best order or approach, even if time is not a constraint. Classically in compilers this is known as the phase ordering problem and various schemes have been proposed to try and find good sequences. Some examples from the literature:
* [Adaptive Optimizing Compilers for the 21st Century](https://www.cs.rice.edu/~keith/EMBED/lacsi2001.pdf)
* [Practical Exhaustive Optimization Phase Order Exploration and Evaluation](http://www.cs.fsu.edu/~whalley/papers/taco09.pdf) 
* [MiCOMP: Mitigating the Compiler Phase-Ordering Problem Using Optimization Sub-Sequences and Machine Learning](https://dl.acm.org/doi/10.1145/3124452)

Reply:
Interesting, thanks glad I asked!